File: apps/api/app/schemas/flights.py
"""
Public-facing flight schemas — what the API accepts and returns.
"""
from typing import Literal
from pydantic import BaseModel, Field


# ─── Search Request ───────────────────────────────────────

class PassengerCount(BaseModel):
    adults: int = Field(default=1, ge=1, le=9)
    children: int = Field(default=0, ge=0, le=9)
    infants: int = Field(default=0, ge=0, le=4)


class FlightSearchRequest(BaseModel):
    origin: str = Field(..., min_length=3, max_length=3, description="IATA airport code")
    destination: str = Field(..., min_length=3, max_length=3, description="IATA airport code")
    departure_date: str = Field(..., description="YYYY-MM-DD")
    return_date: str | None = Field(default=None, description="YYYY-MM-DD for round-trip")
    passengers: PassengerCount = Field(default_factory=PassengerCount)
    cabin: Literal["ECONOMY", "PREMIUM_ECONOMY", "BUSINESS", "FIRST"] = "ECONOMY"


# ─── Search Response ──────────────────────────────────────

class SegmentResponse(BaseModel):
    origin: str
    destination: str
    departure_datetime: str
    arrival_datetime: str
    carrier: str
    flight_number: str
    cabin: str
    duration_minutes: int
    stops: int


class FlightOfferResponse(BaseModel):
    offer_id: str
    price_total: float
    currency: str
    segments: list[SegmentResponse]
    refundable: bool


class FlightSearchResponse(BaseModel):
    search_id: str
    offers: list[FlightOfferResponse]
    total: int


# ─── Booking Request ──────────────────────────────────────

class PassengerDetail(BaseModel):
    type: Literal["ADT", "CHD", "INF"] = "ADT"
    first_name: str
    last_name: str
    date_of_birth: str = Field(..., description="YYYY-MM-DD")
    nationality: str = Field(..., min_length=2, max_length=2, description="ISO 2-letter country code")
    passport_number: str | None = None
    passport_expiry: str | None = Field(default=None, description="YYYY-MM-DD")


class FlightBookingRequest(BaseModel):
    offer_id: str
    passengers: list[PassengerDetail] = Field(..., min_length=1)
    contact_email: str
    contact_phone: str


# ─── Booking Response ─────────────────────────────────────

class FlightBookingResponse(BaseModel):
    booking_reference: str
    status: str
    price_total: float
    currency: str

File: apps/web/lib/api/flights.ts
import { apiRequest } from "./client";

// ─── Types ────────────────────────────────────────────────

export interface PassengerCount {
  adults: number;
  children: number;
  infants: number;
}

export interface FlightSearchParams {
  origin: string;
  destination: string;
  departure_date: string;
  return_date?: string;
  passengers: PassengerCount;
  cabin: "ECONOMY" | "PREMIUM_ECONOMY" | "BUSINESS" | "FIRST";
}

export interface Segment {
  origin: string;
  destination: string;
  departure_datetime: string;
  arrival_datetime: string;
  carrier: string;
  flight_number: string;
  cabin: string;
  duration_minutes: number;
  stops: number;
}

export interface FlightOffer {
  offer_id: string;
  price_total: number;
  currency: string;
  segments: Segment[];
  refundable: boolean;
}

export interface FlightSearchResponse {
  search_id: string;
  offers: FlightOffer[];
  total: number;
}

export interface PassengerDetail {
  type: "ADT" | "CHD" | "INF";
  first_name: string;
  last_name: string;
  date_of_birth: string;
  nationality: string;
  passport_number?: string;
  passport_expiry?: string;
}

export interface BookingRequest {
  offer_id: string;
  passengers: PassengerDetail[];
  contact_email: string;
  contact_phone: string;
}

export interface BookingResponse {
  booking_reference: string;
  status: string;
  price_total: number;
  currency: string;
}

// ─── Cabin label map ──────────────────────────────────────

export const CABIN_MAP: Record<string, FlightSearchParams["cabin"]> = {
  Economy: "ECONOMY",
  "Premium Economy": "PREMIUM_ECONOMY",
  Business: "BUSINESS",
  First: "FIRST",
};

// ─── API calls ────────────────────────────────────────────

export const flightsApi = {
  search: (params: FlightSearchParams) =>
    apiRequest<FlightSearchResponse>("/flights/search", {
      method: "POST",
      body: JSON.stringify(params),
    }),

  book: (params: BookingRequest) =>
    apiRequest<BookingResponse>("/flights/book", {
      method: "POST",
      body: JSON.stringify(params),
    }),

  getBooking: (pnr: string) =>
    apiRequest<Record<string, unknown>>(`/flights/booking/${pnr}`),

  cancelBooking: (pnr: string) =>
    apiRequest<Record<string, unknown>>(`/flights/booking/${pnr}`, {
      method: "DELETE",
    }),
};

// ─── Helpers ──────────────────────────────────────────────

export function formatDuration(minutes: number): string {
  const h = Math.floor(minutes / 60);
  const m = minutes % 60;
  return h > 0 ? `${h}h ${m}m` : `${m}m`;
}

export function formatDateTime(dt: string): { date: string; time: string } {
  if (!dt) return { date: "", time: "" };
  const d = new Date(dt);
  return {
    date: d.toLocaleDateString("en-GB", { day: "2-digit", month: "short" }),
    time: d.toLocaleTimeString("en-GB", { hour: "2-digit", minute: "2-digit" }),
  };
}
